codesearch/schema/schema.go

```diff
@@ -32,49 +20,160 @@ const (
+		NewFieldSchemaOrPanic(types.KeywordField, IDField, false),
+		NewFieldSchemaOrPanic(types.TrigramField, FilenameField, true),
+		NewFieldSchemaOrPanic(types.SparseNgramField, ContentField, true),
+		NewFieldSchemaOrPanic(types.KeywordField, LanguageField, true),
+		NewFieldSchemaOrPanic(types.KeywordField, OwnerField, true),
+		NewFieldSchemaOrPanic(types.KeywordField, RepoField, true),
+		NewFieldSchemaOrPanic(types.KeywordField, SHAField, true),
+	})
 )
 
-// TODO(tylerw): this should come from a flag?
-var skipMime = regexp.MustCompile(`^audio/.*|video/.*|image/.*|application/gzip$`)
+// Satisfies the types.FieldSchema interface.
+type fieldSchema struct {
+	ftype     types.FieldType
+	name      string
+	stored    bool
+	tokenizer types.Tokenizer
+}
+
+func (f fieldSchema) Type() types.FieldType      { return f.ftype }
+func (f fieldSchema) Name() string               { return f.name }
+func (f fieldSchema) Stored() bool               { return f.stored }
+func (f fieldSchema) Tokenizer() types.Tokenizer { return f.tokenizer }
 
-func MakeDocument(name, commitSha string, repoURL *git.RepoURL, buf []byte) (types.Document, error) {
-	// Skip long files.
-	if len(buf) > maxFileLen {
-		return nil, fmt.Errorf("skipping %s (file too long)", name)
+func (f fieldSchema) MakeField(buf []byte) types.Field {
+	return documentField{
+		buf:    buf,
+		schema: f,
 	}
+}
 
-	var shortBuf []byte
-	if len(buf) > detectionBufferSize {
-		shortBuf = buf[:detectionBufferSize]
-	} else {
-		shortBuf = buf
+func NewFieldSchema(ftype types.FieldType, name string, stored bool) (types.FieldSchema, error) {
+	var tokenizer types.Tokenizer = nil
+
+	if !fieldNameRegex.MatchString(name) {
+		return nil, status.InvalidArgumentErrorf("Invalid field name %q", name)
 	}
 
-	// Check the mimetype and skip if bad.
-	mtype, err := mimetype.DetectReader(bytes.NewReader(shortBuf))
-	if err == nil && skipMime.MatchString(mtype.String()) {
-		return nil, fmt.Errorf("skipping %s (invalid mime type: %q)", name, mtype.String())
+	switch ftype {
+	case types.SparseNgramField:
+		// TODO(jdelfino): old code used lowercase on query side but not on index side - was it a bug?
+		tokenizer = token.NewSparseNgramTokenizer(token.WithMaxNgramLength(6), token.WithLowerCase(true))
+	case types.TrigramField:
+		tokenizer = token.NewTrigramTokenizer()
+	case types.KeywordField:
+		tokenizer = token.NewWhitespaceTokenizer()
+	default:
+		return nil, status.InvalidArgumentErrorf("Invalid field type %q", ftype)
 	}
 
-	// Skip non-utf8 encoded files.
-	if !utf8.Valid(buf) {
-		return nil, fmt.Errorf("skipping %s (non-utf8 content)", name)
+	return fieldSchema{
+		ftype:     ftype,
+		name:      name,
+		stored:    stored,
+		tokenizer: tokenizer,
+	}, nil
+}
+
+// Schemas are typically static and const, so in the most common situations, it's reasonable to
+// panic if creation fails.
+func NewFieldSchemaOrPanic(ftype types.FieldType, name string, stored bool) types.FieldSchema {
```

Comment:
probably could call this MustSchemaField or something to match the stdlib

Reply:
👍, forgot about that handy naming convention